### PR TITLE
mrc-2276: Add workflow missing-dependencies endpoint

### PR DESCRIPTION
**Pull request by r-ash**
This PR will:
* Add endpoint to get missing dependencies of a workflow

This is just considering a report as a `dependency` if it is linked with `id: latest` or search term dependencies `id: latest(parameter:nmin == nmin)`. 

There is a bit of an awkward edge case with search terms,
if we have
```
steps:
  - preprocess:
      params:
        nmin: 5
  - process:
      params:
        nmin: 7
  - postprocess:
```
and `process` has `orderly.yml` containing
```
parameters:
  nmin: ~
depends:
  preprocess:
    id: latest(parameter:nmin == nmin)
```

Then should this flag `preprocess` as a missing dependency of `process` because the values of the params are different? I think probably so right? Cautious not to make this validation too clever though, just presenting the information might be easier

Note I am expecting the run workflow to have the same input json (probably with a couple of extra fields but the `tasks` as they are here)

### CODE REVIEW

File: R/workflows.R [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    if (!(task %in% reports)) {
+      stop(sprintf("Task with name '%s' cannot be found.", task))
+    }
+    get_missing_dependencies(task, path, tasks)
+  })
+  list(missing_dependencies = setNames(missing_deps, names(tasks)))
+}
+
+get_missing_dependencies <- function(task, path, tasks) {
+  ## Not ideal that we build a graph of all dependencies then throw all
```
Comment (r-ash) [inline R/workflows.R:13]: I thought `max_depth` would achieve this for us but that just throws an error when that depth is reached. What do you think of adding another arg or changing the `max_depth` arg to return the tree just to that level of depth?

File: R/workflows.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  ## these away deeper than 1 level, should probably add support for that
+  ## in orderly to avoid unnecessary computation
+  graph <- orderly::orderly_graph(task, root = path, direction = "upstream",
+                                  use = "src")
+  dependencies <- lapply(graph$root$children, function(vertex) {
+    scalar(vertex$name)
+  })
+  ## TODO: consider search query dependencies i.e.
```
Comment (r-ash) [inline R/workflows.R:21]: At the moment the graph only returns `latest` dependencies, we want to also depend on if the yml contains something like
```
parameters:
  nmin: ~
depends:
  preprocess:
    id: latest(nmin == nmin)
```

File: R/workflows.R (hunk 1 above)
Comment (r-ash) [reply R/workflows.R]: Added a PR in orderly https://github.com/vimc/orderly/pull/278 to handle this and update code above to use `max_depth = 1`

File: R/workflows.R (hunk 2 above)
Comment (r-ash) [reply R/workflows.R]: I added a test of this in orderly, orderly will return for `latest` dependencies and `latest` with some search parameter so it already does what we want. See https://github.com/vimc/orderly/pull/278/files#r612514715

**Comment by r-ash** [comment]
> Then should this flag preprocess as a missing dependency of process because the values of the params are different? I think probably so right? Cautious not to make this validation too clever though, just presenting the information might be easier

Discussed this and actually let's just return this info to the user and they can choose to include the report or not, and avoid having our code being "too" clever

File: R/api.R [hunk 3]
```diff
@@ -411,6 +412,24 @@ endpoint_run_metadata <- function(runner) {
  )
 }
 
+target_workflow_missing_dependencies <- function(path, body) {
+  body <- jsonlite::fromJSON(body)
+  tasks <- body$tasks
+  workflow_missing_dependencies(path, tasks)
+}
+
+endpoint_workflow_missing_dependencies <- function(path) {
```
Comment (richfitz) [inline R/api.R:421]: What is your thoughts here of making this a function of `path` rather than a method of the runner?

**Review by richfitz — CHANGES_REQUESTED** [review]

File: R/api.R (hunk 3 above)
Comment (r-ash) [reply R/api.R:421]: Trying to stop `runner` becoming some massive object which does everything only going through the `runner` when the function interacts with the queue.

**Review by richfitz — APPROVED** [review]

**Pull request merged**